dali/operators/math/expressions/expression_factory_instances/expression_impl_factory.h

```diff
@@ -86,22 +86,28 @@ std::unique_ptr<ExprImplBase> ExprImplFactoryBinOp(const ExprFunc &expr) {
   std::unique_ptr<ExprImplBase> result;
   auto left_type = expr[0].GetTypeId();
   auto right_type = expr[1].GetTypeId();
+  auto is_non_scalar = [](const ExprNode& node) {
+    return node.GetNodeType() == NodeType::Tensor && !IsScalarLike(node);
+  };
+  auto is_scalar = [](const ExprNode& node) {
+    return IsScalarLike(node);
+  };
   TYPE_SWITCH(left_type, type2id, Left_t, ARITHMETIC_ALLOWED_TYPES, (
     TYPE_SWITCH(right_type, type2id, Right_t, ARITHMETIC_ALLOWED_TYPES, (
       using Out_t = typename arithm_meta<op, Backend>::template result_t<Left_t, Right_t>;
-      if (expr[0].GetNodeType() == NodeType::Tensor && IsScalarLike(expr[1])) {
+      if (is_non_scalar(expr[0]) && is_scalar(expr[1])) {
         result.reset(new ImplTensorConstant<op, Out_t, Left_t, Right_t>());
-      } else if (IsScalarLike(expr[0]) && expr[1].GetNodeType() == NodeType::Tensor) {
-        result.reset( new ImplConstantTensor<op, Out_t, Left_t, Right_t>());
-      } else if (expr[0].GetNodeType() == NodeType::Tensor &&
-                 expr[1].GetNodeType() == NodeType::Tensor) {
+      } else if (is_scalar(expr[0]) && is_non_scalar(expr[1])) {
+        result.reset(new ImplConstantTensor<op, Out_t, Left_t, Right_t>());
+      } else if (is_non_scalar(expr[0]) && is_non_scalar(expr[1])) {
         // Both are non-scalar tensors
         result.reset(new ImplTensorTensor<op, Out_t, Left_t, Right_t>());
       } else {
-        DALI_FAIL("Expression cannot have two scalar operands");
+        // TODO(janton) : do something else?
+        result.reset(new ImplTensorConstant<op, Out_t, Left_t, Right_t>());
```

Comment:
What is this about?

Reply:
There can be shapes that are both reduced to scalar:
{1, 1, 1} an {1}.

Since we don't have a Scalar-Scalar implementation, we just launch the Tensor-Scalar implementation in such case.